NSrtm.Core/Pgm/PgmDataDescription.cs

```diff
@@ -0,0 +1,186 @@
+            {
+                throw new ArgumentOutOfRangeException("rawData");
+            }
+            return rawData * pgmParameters.Scale + pgmParameters.Offset;
+        }
+    }
+
+    internal static class PgmDataDescriptionExtractor
+    {
+        private static readonly Regex LatLonParser = new Regex("^(?<deg>[-+0-9]+)+(?<pos>[ENSW]*)$");
+
+        public static PgmDataDescription FromStream(Stream stream)
+        {
+            var preamble = extractPreambleFromStream(stream);
+            return getConstatantsFromPreamble(preamble);
+        }
+
+        private static string extractPreambleFromStream(Stream stream)
+        {
+            using (var reader = new StreamReader(stream, Encoding.UTF8, false, 1024, true))
+            {
+                const string _pgmMagicNumber = "P5";
+                var preamble = reader.ReadLine();
+                if (preamble != _pgmMagicNumber)
+                {
+                    throw new FileFormatException(String.Format("Wrong portable gray map magic number. Actual {0}, but should be {1}.", preamble, _pgmMagicNumber));
+                }
+
+                string lastLine;
+                do
+                {
+                    lastLine = reader.ReadLine();
+                    preamble += lastLine;
+                    preamble += "\n";
+                } while (lastLine.StartsWith("#", StringComparison.CurrentCulture));
+                preamble += reader.ReadLine();
+
+                return preamble;
+            }
+        }
+
+        private static PgmDataDescription getConstatantsFromPreamble(string preamble) //TODO define magic numbers
+        {
+            var unifyPreamble = preamble.Replace("#", "\n");
+            var words = unifyPreamble.Split(' ', '\n')
+                                     .ToList();
+            var offset = Convert.ToDouble(words[getParameterIndexUsingName(words, "Offset")], CultureInfo.InvariantCulture);
+            var scale = Convert.ToDouble(words[getParameterIndexUsingName(words, "Scale")], CultureInfo.InvariantCulture);
+            var latIndex = getParameterIndexUsingName(words, "Origin");
+            var lat = getLatLonValues(words, latIndex);
+            var lon = getLatLonValues(words, latIndex + 1);
+            var indexOfWidthInPreamble = words.Count - 4;
+            var width = Convert.ToInt32(words[indexOfWidthInPreamble]);
+            int indexOfHightInPreamble = words.Count - 2;
+            var hight = Convert.ToInt32(words[indexOfHightInPreamble]);
+            var maxValue = Convert.ToInt32(words.Last());
+
+            return new PgmDataDescription(offset, scale, lat, lon, width, hight, maxValue, preamble.Length);
+        }
+
+        private static int getParameterIndexUsingName(List<string> words, string parameter)
+        {
+            return words.FindIndex(v => v == parameter) + 1;
+        }
+
+        private static int getLatLonValues(List<string> words, int index)
```

Comment:
Why this function accepts whole array of words?

Why this collection is called `words`?